src/main/java/net/minecraftforge/common/capabilities/AttachCapabilitiesEvent.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.common.capabilities;
+
+
+import net.minecraft.core.Direction;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ * 
+ */
+public abstract class AttachCapabilitiesEvent<T extends ICapabilityProvider> extends Event {
+
+	protected final T obj;
+	protected final AttachedCapability<?>[][] caps = new AttachedCapability<?>[CapabilityManager.getNumCaps()][7];
+	protected int size = 0;
+
+	protected AttachCapabilitiesEvent(T obj) {
+		this.obj = obj;
+		System.out.println("Initialized " + obj.toString() + " hash " + obj.hashCode());
+	}
+
+	/**
+	 * Retrieves the object that is being created, Not much state is set.
+	 */
+	public final T getObject() {
+		return this.obj;
+	}
+
+	public final int getSize() {
+		return this.size;
+	}
+	
+	/**
+	 * Adds a capability to be attached to this object. Keys MUST be unique, it is
+	 * suggested that you set the domain to your mod ID. If the capability is an
+	 * instance of INBTSerializable, this key will be used when serializing this
+	 * capability.
+	 *
+	 * @param key The name of owner of this capability provider.
+	 * @param cap The capability provider
+	 */
+	public <C> AttachedCapability<C> addCapability(ResourceLocation id, Capability<C> cap, LazyOptional<C> inst,
+			Direction... sides) {
+		if (!this.obj.isDirectionSensitive())
+			sides = new Direction[] { null };
+		AttachedCapability<C> attached = new AttachedCapability<>(id, cap, inst, sides);
+		for (Direction side : sides) {
+			int dKey = side == null ? 6 : side.ordinal();
+			if (caps[cap.getId()][dKey] == null) {
+				caps[cap.getId()][dKey] = attached;
+				this.size++;
+			} else {
+				throw new RuntimeException("Duplicated capabilities!"); // TODO: Change to a better message.
```

Comment:
I was struggling to figure out the best method of what to do here - because this method has to return the AttachedCapability instance unless we want the consumer to make that instance on their own.

If we did it that way, we could have this method return a Direction[] which has the sides it was successfully attached to.

Reply:
1. Decide which wins: newly attached capabilities or preexisting ones (I'd lean towards preexisting ones), however having newly attached capabilities would make the implementation a tad cleaner.
2. Either silently overwrite, or silently ignore, as per (1).
3. Return the `AttachedCapability` instance if *at least one* side was attached, otherwise return `null`. If preexisting capabilities win, first filter the `sides` as to only comprise sides that were actually attached.

This seems to me like it solves all the stated problems:

- The caller receives their attached capability instance
- The attached capability's `sides` / `getSides()` can be used to see what sides it *actually* attached to.
- A return value of `null` indicates that the capability didn't attach to anything.

Reply:
The function is now updated to prefer existing caps over overwriting old ones.  Ideally nobody ever has conflicts, but we can't guarantee that in practice.